src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -1142,55 +1146,67 @@ impl LinkCollector<'_, '_> {
+                // These are namespaces; allow anything in the namespace to match
+                | (_, Some(Disambiguator::Namespace(_)))
+                // If no disambiguator given, allow anything
+                | (_, None)
+                // All of these are valid, so do nothing
+                => {}
+                (actual, Some(Disambiguator::Kind(expected))) if actual == expected => {}
+                (_, Some(specified @ Disambiguator::Kind(_) | specified @ Disambiguator::Primitive)) => {
+                    report_mismatch(specified, Disambiguator::Kind(kind));
+                    return None;
                 }
+            }
+
+            // item can be non-local e.g. when using #[doc(primitive = "pointer")]
+            if let Some((src_id, dst_id)) = id
+                .as_local()
+                .and_then(|dst_id| item.def_id.as_local().map(|src_id| (src_id, dst_id)))
+            {
+                use rustc_hir::def_id::LOCAL_CRATE;
 
-                // item can be non-local e.g. when using #[doc(primitive = "pointer")]
-                if let Some((src_id, dst_id)) = id
-                    .as_local()
-                    .and_then(|dst_id| item.def_id.as_local().map(|src_id| (src_id, dst_id)))
+                let hir_src = self.cx.tcx.hir().local_def_id_to_hir_id(src_id);
+                let hir_dst = self.cx.tcx.hir().local_def_id_to_hir_id(dst_id);
+
+                if self.cx.tcx.privacy_access_levels(LOCAL_CRATE).is_exported(hir_src)
+                    && !self.cx.tcx.privacy_access_levels(LOCAL_CRATE).is_exported(hir_dst)
                 {
-                    use rustc_hir::def_id::LOCAL_CRATE;
+                    privacy_error(cx, &item, &path_str, dox, &ori_link);
+                }
+            }
 
-                    let hir_src = self.cx.tcx.hir().local_def_id_to_hir_id(src_id);
-                    let hir_dst = self.cx.tcx.hir().local_def_id_to_hir_id(dst_id);
+            Some((kind, id))
+        };
 
-                    if self.cx.tcx.privacy_access_levels(LOCAL_CRATE).is_exported(hir_src)
-                        && !self.cx.tcx.privacy_access_levels(LOCAL_CRATE).is_exported(hir_dst)
-                    {
-                        privacy_error(cx, &item, &path_str, dox, &ori_link);
+        match res {
+            Res::Primitive(_) => {
+                if let Some((kind, id)) = self.kind_side_channel.take() {
+                    verify(kind, id)?;
+                } else {
+                    match disambiguator {
+                        Some(Disambiguator::Primitive | Disambiguator::Namespace(_)) | None => {}
+                        Some(other) => {
+                            report_mismatch(other, Disambiguator::Primitive);
+                            return None;
+                        }
                     }
                 }
+                Some(ItemLink { link: ori_link.link, link_text, did: None, fragment })
```

Comment:
This doesn't look right for the verify case: I think `register_res` is still needed, and `did` should be the ID you just found. (I think register_res only matters for cross-crate re-exports, which is why it didn't show up in the tests.)

Maybe it would be simpler rather than duplicating the control flow of `Res::Def` to do `let (kind, id) = match res { ... }`, like you had in https://github.com/rust-lang/rust/pull/80660/commits/fc3a4058cefc09bebe98039c7fb8d74d1be9c6e6 ? That looked right to me, I'm not sure why you changed it.

Reply:
Including the `let id = register_res(cx, rustc_hir::def::Res::Def(kind, id));` causes this error:
```
error: internal compiler error: compiler/rustc_metadata/src/rmeta/decoder.rs:807:18: def-index does not refer to trait or trait alias

thread 'rustc' panicked at 'Box<Any>', compiler/rustc_errors/src/lib.rs:958:9
```
I changed the control flow to avoid it. Should I be passing `register_res` something else?

Reply:
Well that's strange. Can you find where that goes wrong? I wouldn't expect a fake id to make it this far.

If this ends up being tricky I'm ok with getting the fix in, but it sounds like something suspicious is going on.

It would also be nice to add a test for cross crate exports:

```rust
//! [trim]
pub use str::trim;
```

Reply:
`use`ing primitive associated items doesn't appear to work in the first place ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=1208a24fa33894a880b22505ad7a21ec))

Reply:
Hmm, ok. This is probably ok to land as-is then - I'm confused why `register_res` ICEs but otherwise the current code seems fine.

Can you leave a comment summarizing what's going on?